src/crush/mapper.c

```diff
@@ -319,40 +326,51 @@ static inline __s32 *get_choose_arg_ids(const struct crush_bucket_straw2 *bucket
   return arg->ids;
 }
 
+/*
+ * Compute exponential random variable using inversion method.
+ *
+ * for reference, see the exponential distribution example at:  
+ * https://en.wikipedia.org/wiki/Inverse_transform_sampling#Examples
+ */
+static inline __s64 generate_exponential_distribution(int type, int x, int y, int z, 
+                                                      int weight)
+{
+  unsigned int u = crush_hash32_3(type, x, y, z);
+  u &= 0xffff;
+
+  /*
+   * for some reason slightly less than 0x10000 produces
+   * a slightly more accurate distribution... probably a
+   * rounding effect.
+   *
+   * the natural log lookup table maps [0,0xffff]
+   * (corresponding to real numbers [1/0x10000, 1] to
+   * [0, 0xffffffffffff] (corresponding to real numbers
+   * [-11.090355,0]).
+   */
+  __s64 ln = crush_ln(u) - 0x1000000000000ll;
+
+  /*
+   * divide by 16.16 fixed-point weight.  note
+   * that the ln value is negative, so a larger
+   * weight means a larger (less negative) value
+   * for draw.
+   */
+  return div64_s64(ln, weight);
```

Comment:
oops, one formatting nit: can you use 8 chars or tab here?  this file is formatted in kernel C style.  thanks!

Reply:
changed